lib/delta-types/src/Data/Delta/Embedding.hs

```diff
@@ -0,0 +1,256 @@
+{-# LANGUAGE BangPatterns #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE TupleSections #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE TypeOperators #-}
+
+{-|
+Copyright: © 2021-2023 IOHK, 2024 Cardano Foundation
+License: Apache-2.0
+
+Embeddings of delta types.
+-}
+module Data.Delta.Embedding (
+    -- $Embedding
+    -- * Embedding
+      Embedding
+    , module Data.Semigroupoid
+    , Embedding' (..)
+    , mkEmbedding
+    , fromEmbedding
+    , pair
+    , liftUpdates
+    , replaceFromApply
+
+    -- * Internal
+    , inject
+    , project
+    , Machine (..)
+    , idle
+    , pairMachine
+    , fromState
+    ) where
+
+import Prelude
+
+import Control.Exception
+    ( SomeException
+    )
+import Data.Delta.Core
+    ( Delta (..)
+    , Replace (..)
+    )
+import Data.Either
+    ( fromRight
+    )
+import Data.Semigroupoid
+    ( Semigroupoid (..)
+    )
+
+{-------------------------------------------------------------------------------
+    Embedding
+-------------------------------------------------------------------------------}
+{- $Embedding
+
+An 'Embedding'@ da db@ embeds one type and its delta type @da@
+into another type and its delta type @db@.
+
+For reasons of efficiency, 'Embedding' is an abstract type.
+It is constructed using the 'Embedding'' type, which has
+three components.
+
+* 'write' embeds values from the type @a = 'Base' da@
+    into the type @b = 'Base' db@.
+* 'load' attempts to retrieve the value of type @a@
+    from the type @b@, but does not necessarily succeed.
+* 'update' maps a delta type @da@ to a delta type @db@.
+    For this mapping, the value of type @a@ and a corresponding
+    value of type @b@ are provided;
+    the delta types @da@ and @db@ are
+    relative to these values.
+
+The embedding of one type into the other is characterized by the following
+properties:
+
+* The embedding need __not__ be __surjective__:
+    The type @b@ may contain many values that do not correspond to
+    a value of type @a@. Hence, 'load' has an 'Either' result.
+    (See Note [EitherSomeException] for the choice of exception type.)
+    However, retrieving a written value always succeeds, we have
+
+        > load . write = Right
+
+* The embedding is __redundant__:
+    The type @b@ may contain multiple values that correspond to
+    one and the same @a@.
+    This is why the 'update' function expects the type @b@ as argument,
+    so that the right deltas can be computed.
+    Put differently, we often have
+
+        > write a ≠ b   where Right a = load b
+
+* The embedding of a delta __commutes with 'apply'__.
+    We have
+
+        > Just (apply da a) = load (apply (update a b da) b)
```

Comment:
you missed `prop` to be consistent with `Delta.Core` , correct?

Reply:
Not using `prop>` here, because Haddock doesn't support multiline `prop>` statements.